### PR TITLE
SQLite Custom Functions are now supported

### PR DESCRIPTION
**Pull request by danramteke**
This PR adds custom function support to `sqlite-nio`. Both custom functions and custom aggregates are supported. To find out more about these features, please see the SQLite website about them: https://www.sqlite.org/appfunc.html

There are some queries only possible using custom functions. And further, custom function help soften the need for custom builds of SQLite with various parts of the SQLite amalgamation enabled or disabled: one can simply write a custom function instead.

I have adapted this code from `https://github.com/groue/GRDB.swift/blob/master/GRDB/Core/DatabaseFunction.swift` by @groue. Because GRDB doesn't support Linux or NIO, I was motivated to adapt his code.

Example:
```swift
func testFunctionArgumentString() throws {
	let conn = try SQLiteConnection.open(storage: .memory, threadPool: self.threadPool, on: self.eventLoop).wait()
	defer { try! conn.close().wait() }
	let fn = SQLiteCustomFunction("f", argumentCount: 1) { (values: [SQLiteData]) in
		return values[0].string
	}
	try conn.install(customFunction: fn).wait()
	XCTAssertNil(try conn.query("SELECT f(NULL) as result")
								.map { rows in rows[0].column("result")?.string }.wait())
	XCTAssertEqual("foo", try conn.query("SELECT f('foo') as result")
									.map { rows in rows[0].column("result")?.string }.wait())
}
```

There are many more examples in the new unit test file.

Lastly, is there a [swift-format](https://github.com/apple/swift-format) configuration I can use to format my PR to match the Vapor's code style?

### CODE REVIEW

File: .github/workflows/test.yml
Comment (0xTim) [inline .github/workflows/test.yml:13]: This needs to stay as 5.2 to ensure we don't break 5.2 compatibility

**Review by 0xTim — COMMENTED** [review]
Initial scan looks good, I'll do a proper review next week. Note the comment about CI

File: .github/workflows/test.yml
Comment (danramteke) [reply .github/workflows/test.yml]: Fixed! 

I updated the `maxim-lobanov/setup-xcode` to `v1.2.1` since macOS build won't start.

Comment (gwynne) [reply .github/workflows/test.yml:13]: @0xTim 5.2 compatibility is checked by the main Linux workflows in each repo (at least, that's how it's supposed to work) - I prefer to run the dependent checks in a "current"-version container to reflect the most common usage, since those checks don't have a version matrix.

**Comment by groue** [comment]
> I have adapted this code from `https://github.com/groue/GRDB.swift/blob/master/GRDB/Core/DatabaseFunction.swift` by @groue.

Hello, and thank you for getting inspired from my work 👍 

GRDB uses the the MIT license. It is a permissive license, which allows you to do anything you like with the licensed code. The only requirement is:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

The SQLiteCustomFunction.swift file in this PR contains substantial portions of https://github.com/groue/GRDB.swift/blob/v5.8.0/GRDB/Core/DatabaseFunction.swift: please include the GRDB license in your derived work.

**Comment by danramteke** [comment]
Any updates? @gwynne  @0xTim

File: Sources/SQLiteNIO/SQLiteConnection.swift
Comment (0xTim) [inline Sources/SQLiteNIO/SQLiteConnection.swift:196]: These should probably be trace logs
```suggestion
		logger.trace("Adding custom function \(customFunction.name)")
```

File: Sources/SQLiteNIO/SQLiteConnection.swift
Comment (0xTim) [inline Sources/SQLiteNIO/SQLiteConnection.swift:210]: Trace as above 
```suggestion
		logger.trace("Removing custom function \(customFunction.name)")
```

**Review by 0xTim — CHANGES_REQUESTED** [review]
LGTM other than the log levels. I'll let @gwynne have a look before merging though

**Comment by danramteke** [comment]
Wonderful. I committed both of your suggested changes, @0xTim.

**Comment by Austinpayne** [comment]
@danramteke Can I help get this across the finish line? I am interested in this functionality as well. I'd be happy to fill out the NOTICES.txt if you provided me access to your branch.

**Comment by danramteke** [comment]
@Austinpayne had no idea anyone else wanted this, thanks for the nudge!

**Comment by danramteke** [comment]
@0xTim I did my best to add a `NOTICES.txt` as requested.

**Comment by Austinpayne** [comment]
> @Austinpayne had no idea anyone else wanted this, thanks for the nudge!

Thank you for implementing it! This unlocks some very interesting use cases and I actually had started on an implementation before seeing yours. Yours in much more feature rich so really looking forward to seeing in mainlined.

**Review by 0xTim — APPROVED** [review]
Thanks!

File: Sources/SQLiteNIO/SQLiteConnection.swift
Comment (gwynne) [inline Sources/SQLiteNIO/SQLiteConnection.swift:214]: Per [the `NIOThreadPool` inline docs](https://github.com/apple/swift-nio/blob/main/Sources/NIOPosix/NIOThreadPool.swift#L130), use `runIfActive(eventLoop:_:)` instead:
```suggestion
        return self.threadPool.runIfActive(eventLoop: self.eventLoop) {
            try customFunction.install(in: self)
        }
```

This needs to be fixed pretty much everywhere in this file - this PR is an excellent opportunity to do so 🙂.

File: Sources/SQLiteNIO/SQLiteConnection.swift
Comment (gwynne) [inline Sources/SQLiteNIO/SQLiteConnection.swift:228]: `runIfActive()` again

File: Sources/SQLiteNIO/SQLiteDataConvertible.swift
Comment (gwynne) [inline Sources/SQLiteNIO/SQLiteDataConvertible.swift:54]: Adding this conformance will cause infinite recursion problems in SQLiteKit data marshaling logic. The easiest alternative is usually to special-case values of `SQLiteData` type at the point of use.

File: Sources/SQLiteNIO/SQLiteError.swift
Comment (gwynne) [inline Sources/SQLiteNIO/SQLiteError.swift:128]: About time this missing conversion was added 😆 But please fix the excessive indent 🤣

File: Tests/SQLiteNIOTests/SQLiteCustomFunctionTests.swift
Comment (gwynne) [inline Tests/SQLiteNIOTests/SQLiteCustomFunctionTests.swift:367]: Please use `1` as per the existing tests, or if multiple threads are strictly required, `System.coreCount`.

**Review by gwynne — CHANGES_REQUESTED** [review]
I've commented inline about several minor issues. I'm also very uncomfortable with the amount of manual memory management being thrown around; I'll try to persuade @Lukasa to take a look at what's going on in that regard.

That concern aside, the overall code seems pretty solid and there's no question that having this functionality is both useful and desirable; if we can address the above, this will be good to go!

File: Tests/SQLiteNIOTests/SQLiteCustomFunctionTests.swift
Comment (danramteke) [reply Tests/SQLiteNIOTests/SQLiteCustomFunctionTests.swift]: Done

File: Sources/SQLiteNIO/SQLiteError.swift
Comment (danramteke) [reply Sources/SQLiteNIO/SQLiteError.swift]: Fixed using `Ctrl-i` to re-indent.

File: Sources/SQLiteNIO/SQLiteDataConvertible.swift
Comment (danramteke) [reply Sources/SQLiteNIO/SQLiteDataConvertible.swift]: Fixed

File: Sources/SQLiteNIO/SQLiteConnection.swift
Comment (danramteke) [reply Sources/SQLiteNIO/SQLiteConnection.swift]: Fixed. No more uses of `threadPool.submit` in the file

File: Sources/SQLiteNIO/SQLiteCustomFunction.swift
Comment (danramteke) [reply Sources/SQLiteNIO/SQLiteCustomFunction.swift]: done

File: Sources/SQLiteNIO/SQLiteCustomFunction.swift
Comment (danramteke) [reply Sources/SQLiteNIO/SQLiteCustomFunction.swift]: done

File: Sources/SQLiteNIO/SQLiteCustomFunction.swift
Comment (danramteke) [reply Sources/SQLiteNIO/SQLiteCustomFunction.swift:303]: I understand what the reviewer is saying. I'm open to help for re-implementing this function 💜

**Comment by Austinpayne** [comment]
@danramteke I'm confused, you appear to have reverted the changes of https://github.com/vapor/sqlite-nio/pull/35 in this PR. Was that intentional?

**Comment by gwynne** [comment]
That's due to this branch not having been updated in a while. The conflicts need to be resolved; the seeming "revert" is just an artifact of that. @danramteke, can you get this up to date with the changes from `main`?

**Comment by gwynne** [comment]
And also @Lukasa could you take a peek and let us know if your issues were resolved? 🙂

**Comment by danramteke** [comment]
@Austinpayne yep, @gwynne is right, it's the result of merge conflicts

**Comment by danramteke** [comment]
If anyone can help get this PR over the finish line, please lend a helping hand, as my bandwidth is limited at the moment. 

Remaining tasks:
- [ ] This comment https://github.com/vapor/sqlite-nio/pull/25#discussion_r1039354186
- [ ] updating with changes on the `main` branch

**Comment by gwynne** [comment]
@danramteke @Lukasa Okay - I rebased onto `main` (as it is in your fork) to get rid of the mixed merge commits (merging from upstream main was getting messed up by the merge revert on the branch), fixed up the whitespace confusion, merged upstream main in so everything's completely up to date, and most importantly, I reimplemented the `umanagedAggregateContext()` method to - I hope! - address Cory's very valid concerns.

Cory, did I get it even remotely right? 🙏

File: Sources/SQLiteNIO/SQLiteCustomFunction.swift
Comment (gwynne) [reply Sources/SQLiteNIO/SQLiteCustomFunction.swift:293]: I've taken a crack at it, hopefully Cory won't tell me I've made things even worse 😅

**Comment by gwynne** [comment]
P.S.: Yes, Cory, I know I didn't make the method optionally consuming like you wanted, but the behavior of SQLite with regards to aggregate contexts during xFinal is so weird I wasn't entirely good with the idea of tinkering more than necessary 😕

**Comment by gwynne** [comment]
Ugh, GitHub’s UI some days… 😬

**Comment by Lukasa** [comment]
What happened to the indentation in SQLiteCustomFunction.swift? It's all over the place, the code is _extraordinarily_ hard to read.

**Comment by gwynne** [comment]
> What happened to the indentation in SQLiteCustomFunction.swift? It's all over the place, the code is _extraordinarily_ hard to read.

I thought I managed to fix that 😕 Heck with it, I’ll give it a swiftformat pass once I fix the 5.6 build

File: Sources/SQLiteNIO/SQLiteCustomFunction.swift
Comment (Lukasa) [inline Sources/SQLiteNIO/SQLiteCustomFunction.swift:246]: This seems like a roundabout way of spelling `takeRetainedValue`.

File: Sources/SQLiteNIO/SQLiteData.swift
Comment (Lukasa) [inline Sources/SQLiteNIO/SQLiteData.swift:131]: There's a lot happening in the blob case, but we can make it simpler by using some newer NIO features (and omitting the unnecessary pointer binding):

```swift
if let bytes = sqlite_nio_sqlite3_value_blob(sqliteValue) {
    let count = Int(sqlite_nio_sqlite3_value_bytes(sqliteValue))
    let buffer = ByteBuffer(bytes: UnsafeRawBufferPointer(start: bytes, count: count))
    self = .blob(buffer)
}
```

File: Sources/SQLiteNIO/SQLiteCustomFunction.swift
Comment (gwynne) [reply Sources/SQLiteNIO/SQLiteCustomFunction.swift]: I couldn't agree more 😅

**Review by gwynne — APPROVED** [review]

**Comment by gwynne** [comment]
@groue It seems as if I managed to stumble my way into what Cory was looking for in terms of improvements here - please feel free to use it in GRDB!

**Comment by groue** [comment]
> @groue It seems as if I managed to stumble my way into what Cory was looking for in terms of improvements here - please feel free to use it in GRDB!

That's lovely 😊 I certainly follow this issue and your improvements. The SQLite api regarding custom functions and aggregates is awful, let's say "sooooo C" (quite uncommon, SQLite is usually a very nice C library). Thank you very much again.

**Comment by danramteke** [comment]
Thank you @gwynne !!!

**Comment by gwynne** [comment]
Deliberately merging without a semver- label, will release manually.

**Pull request merged**
